### ISSUE TITLE
no-unsafe-call: ban calling Function

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#9108]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Unintuitively, TS narrows `typeof x === 'function'` to the unsafe `Function` type if x is not already constrained. However, it is safe to use on a constrained type. We should report on unconstrained callables.

### Fail Cases

```TypeScript
declare const maybeFunction: unknown;
if (typeof maybeFunction === 'function') {
    // TS allows this.
    maybeFunction('call', 'with', 'any', 'args');
}
```

### Pass Cases

```TypeScript
declare const maybeFunction: string | (() => string);
if (typeof maybeFunction === 'function') {
    // TS errors if more args are provided than the declared call signature, so this is safe.
    maybeFunction();
}
```

### Additional Info

it's probably fine/good to allow `typeof x === 'function` as long as `x` isn't called. I'm not sure exactly where the line should be drawn between reporting and not reporting. It seems like it should be reported if we _pass_ `x` to a function that might expect a specific call signature. Maybe a heuristic could be `typeof x === 'function` is ok as long as `x` isn't referenced in the `true` branch of the conditional?

**Comment by bradzacher** [comment]
This would probably be better as a part of `no-unsafe-call`. Eg ban all calls of a `Function` typed value.

**Comment by kirkwaiblinger** [comment]
I think I'm ok with that, too. After a little experimentation i think that's safe as far as passing the narrowed function, too, since 
```ts
declare const looselyTypedFunction: Function;
// TS error, not assignable.
const strictlyTypedFunction: () => void = looselyTypedFunction;

// ditto for passing as a parameter to functions
declare function callsWithAConcreteCallSignature(f: () => void): void;
// TS error
callsWithAConcreteCallSignature(looselyTypedFunction)
```
So I _think_ `Function` is only unsafe when directly called, which would get caught with `no-unsafe-call`

**Comment by kirkwaiblinger** [comment]
I guess the downside is just that I could see wanting `no-unsafe-call` disabled but still wanting the sneaky `Function` call safety

**Comment by JoshuaKGoldberg** [comment]
💡🧠 completely different direction: what if we instead made a rule like `no-unsafe-function-type` as a part of #9102? It could be aimed at catching any case where the unsafe `Function` type creeps in:

* `: Function` type annotations
* `typeof` type narrowing on an `unknown`

`Function` is already different from the other types banned in #9102's current new `no-uppercase-wrapper-types` in that it's not a class wrapper around a primitive. And thanks to this issue we can see there's this second difference around `typeof` narrowing too. That makes me suspect we should have a dedicated rule for the `Function` type rather than piecemeal it across multiple other somewhat-related rules (`no-uppercase-wrapper-types`, `no-unsafe-call`).

**Comment by bradzacher** [comment]
The problem with doing that is that it requires type info.
`no-uppercase-wrapper-types` will be pure syntactic.

We don't want to gate a `Function` ban behind type-info.

**Comment by Josh-Cena** [comment]
So is our conclusion to build this as a part of `no-unsafe-call` instead? Should this be an option or just enabled by default?

**Comment by kirkwaiblinger** [comment]
My hot take would be to just put both the existing behavior of `no-unsafe-call` and the Function safety under separate options in order to accomodate https://github.com/typescript-eslint/typescript-eslint/issues/9108#issuecomment-2116196990. I don't feel strongly though

**Comment by bradzacher** [comment]
I don't think there's a case where you want to be unsafe in every case. I think no option is fine - people can use a disable comment to opt-out on a case-by-case basis.

**Comment by kirkwaiblinger** [comment]
> I don't think there's a case where you want to be unsafe in every case. What kind of codebase would say "yes I **always** want to allow unsafely calling `Function`"? I think that kind of user wouldn't even use the rule to begin with.

100% agree. 

> I think no option is fine - people can use a disable comment to opt-out on a case-by-case basis.

However, I don't think this conclusion follows. It _does_ make sense to want to allow calling `any` (i.e. people who don't currently use the rule at all) and ban calling `Function`.

**Comment by Josh-Cena** [comment]
I don't see how `Function` is any more or less safe than `any`. You either ban both or neither

**Comment by bradzacher** [comment]
I'd agree. I don't see why you'd ban one without the other. Both are very clear and present vectors for unsafety and they are really the same thing.

**Comment by kirkwaiblinger** [comment]
I agree that there is zero difference in safety. But, there is a difference in ergonomics and developer intention. Banning using `any` as `any` is very unwieldy for when you want to just do some intentional yolo coding, but the `typeof function` laxness is just sneaky unsafety that looks safe, not an intentional opt-out from type safety.

**Comment by bradzacher** [comment]
My thinking would be this: most people are going to just enable the rule with the defaults and do no more - we have learned that the majority of users don't want to read docs and granularly configure things and instead want sensible defaults.

So adding an option would likely be unused as we'd default it to true. Most users wouldn't read the docs and decide they want the rule on but with just the Function check only.

I definitely understand your reasoning - but I think that it's just not going to be a usecase people want (and thus unnecessary branching for us).

I'm not saying we never add an option! If people ask for it then we should. But I think shipping without it is completely okay.

**Comment by kirkwaiblinger** [comment]
> I'm not saying we never add an option! If people ask for it then we should. But I think shipping without it is completely okay.

Ah, gotcha. I fully agree with this! 🙂

**Comment by controversial** [comment]
Is there any way to narrow `unknown` into something callable after this change?

I have an unknown value `schema` which I expect to contain a Zod schema, and I want to call it safely.

I used to do: 
```ts
if (typeof schema !== 'object' || schema === null || !('parse' in schema) || typeof schema.parse !== 'function') throw new Error(`Expected a Zod schema`);
try {
  schema.parse(myData);
}
```

This narrows `schema.parse` to Function, which is no longer allowed to be called. I understand now that calling `Function` is unsafe, but I’m not clear on what the “safe” way is to narrow `unknown` into something callable. 

Maybe the docs for the `no-unsafe-call` rule could address this?

**Comment by kirkwaiblinger** [comment]
@controversial Interesting question. JS doesn't really provide any way to know something is safely callable at runtime, so I'm not sure of a "safe" way to narrow `unknown` into something callable by runtime checks.

Consider 
```js
class Foo {};

typeof Foo; // returns 'function'
Foo.length; // returns 0

// so I should be safe to call this with 0 args, right?

Foo(); // Uncaught TypeError: Class constructor Foo cannot be invoked without 'new'
```

With that in mind, the following will pass your checks, but still cause a runtime exception
```js
const schema = { parse: Foo }; 
```

I think your best bet is to use eslint-disable comments and/or type assertions and surround stuff in try/catch 🤷

**Comment by JoshuaKGoldberg** [comment]
I think this'd be a good note for the docs, yeah. If anybody wants to file an issue and/or send a PR, that's a 👍 from me!
